pkg/image/strategy/simplelookup.go

```diff
@@ -0,0 +1,226 @@
+	alternates *alternates
+	icspClient operatorv1alpha1client.ImageContentSourcePolicyInterface
+}
+
+func NewSimpleLookupICSPStrategy(file string, client operatorv1alpha1client.ImageContentSourcePolicyInterface) registryclient.AlternateRepositoryStrategy {
+	return &simpleLookupICSP{
+		icspFile:   file,
+		icspClient: client,
+		loaded:     false,
+	}
+}
+
+func (s *simpleLookupICSP) FirstRequest(ctx context.Context, locator reference.DockerImageReference) (alternateRepositories []reference.DockerImageReference, err error) {
+	return nil, nil
+}
+
+// WithAlternateSources returns a list of possible image references for the locator image reference, gathered from image content source policies, or an error
+func (s *simpleLookupICSP) WithAlternateSources(ctx context.Context, locator reference.DockerImageReference, reason error) (alternateRepositories []reference.DockerImageReference, err error) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	var altList []reference.DockerImageReference
+	if !s.loaded {
+		if err := s.resolve(locator); err != nil {
+			return nil, err
+		}
+	}
+	if len(s.alternates.sources) == 0 {
+		return nil, fmt.Errorf("no alternative image references found for image: %s", locator.String())
+	}
+	for _, alt := range s.alternates.sources {
+		altList = append(altList, alt)
+	}
+	return altList, nil
+}
+
+// RepositoryForAlternateReference returns an appropriate image reference for a given Context and repository. For simple lookup strategy,
+// this is the first from the map, with no other ordering. For more complex ordering, use another strategy.
+func (s *simpleLookupICSP) RepositoryForAlternateReference(ctx context.Context, ref reference.DockerImageReference, cachedRepos map[reference.DockerImageReference]distribution.Repository) (distribution.Repository, reference.DockerImageReference, error) {
+	for altRef, altRepo := range cachedRepos {
+		if altRepo == nil {
+			continue
+		}
+		manifests, err := altRepo.Manifests(ctx)
+		if err != nil {
+			continue
+		}
+		// best effort if passed an ID check it, if not then skip the check and return it
+		// let the caller determine if this is a legitimate ref, since this is what they passed
+		// ie, if it is coming from anywhre internal, it will have an ID, if not, then it's
+		// coming straight from user.
+		if len(ref.ID) != 0 {
+			dgst, err := digest.Parse(ref.ID)
+			if err != nil {
+				continue
+			}
+			_, err = manifests.Get(ctx, dgst)
+			if err != nil {
+				continue
+			}
+		}
+		altRef.ID = ref.ID
+		altRef.Tag = ref.Tag
+		return altRepo, altRef, nil
+	}
+	return nil, ref, fmt.Errorf("did not find alternative image reference for %s", ref.String())
+}
+
+// addICSPsFromCluster will lookup ImageContentSourcePolicy resources in cluster.
+func (s *simpleLookupICSP) addICSPsFromCluster() error {
+	if s.icspClient == nil {
+		return fmt.Errorf("no client to access ImageContentSourcePolicies in cluster")
+	}
+	icsps, err := s.icspClient.List(context.TODO(), metav1.ListOptions{})
```

Comment:
nit: wouldn't it be better to pass down the context from `WithAlternateSources` here?

Reply:
yes, thanks, fixed